tests/Aql/ExecutionBlockImplTest.cpp

```diff
@@ -391,6 +401,1167 @@ TEST_F(ExecutionBlockImplTest,
+  RegisterId outRegSecond = 1;
+  auto producerFirst = produceBlock(singleton.get(), builder, outRegFirst);
+  auto producer = produceBlock(producerFirst.get(), builder, outRegSecond);
+  auto const& call = GetParam();
+  AqlCallStack stack{call};
+  auto const [state, skipped, block] = producer->execute(stack);
+  if (call.getLimit() < 100) {
+    if (call.hasHardLimit()) {
+      // On hard limit we need to stop
+      EXPECT_EQ(state, ExecutionState::DONE);
+    } else {
+      // On soft limit we need to be able to produce more
+      EXPECT_EQ(state, ExecutionState::HASMORE);
+    }
+  } else {
+    EXPECT_FALSE(call.hasHardLimit());
+    EXPECT_EQ(state, ExecutionState::DONE);
+  }
+
+  auto firstRegBuilder = std::make_shared<VPackBuilder>();
+  auto secondRegBuilder = std::make_shared<VPackBuilder>();
+  firstRegBuilder->openArray();
+  secondRegBuilder->openArray();
+  for (size_t i = 0; i < 10; ++i) {
+    // i => 0 -> 9
+    for (size_t j = 0; j < 10; ++j) {
+      // j => 0 -> 9
+      firstRegBuilder->add(VPackValue(i));
+      secondRegBuilder->add(VPackValue(j));
+    }
+  }
+  secondRegBuilder->close();
+  firstRegBuilder->close();
+  ValidateResult(firstRegBuilder, skipped, block, outRegFirst);
+  ValidateResult(secondRegBuilder, skipped, block, outRegSecond);
+}
+
+// Explicitly test call forwarding, on exectors.
+// We use two pass-through producers, that simply copy over input and assert an calls.
+// On top of them we have a 1000 line producer.
+// We expect the result to be identical to the 1000 line producer only.
+TEST_P(ExecutionBlockImplExecuteIntegrationTest, test_call_forwarding_passthrough) {
+  auto singleton = createSingleton();
+
+  auto builder = std::make_shared<VPackBuilder>();
+  builder->openArray();
+  for (size_t i = 0; i < 1000; ++i) {
+    builder->add(VPackValue(i));
+  }
+  builder->close();
+  RegisterId outReg = 0;
+  auto producer = produceBlock(singleton.get(), builder, outReg);
+
+  CallAsserter upperState{GetParam()};
+  CallAsserter lowerState{GetParam()};
+
+  auto testForwarding =
+      [&](AqlItemBlockInputRange& inputRange,
+          OutputAqlItemRow& output) -> std::tuple<ExecutorState, LambdaExe::Stats, AqlCall> {
+    while (inputRange.hasDataRow() && !output.isFull()) {
+      auto const& [state, input] = inputRange.nextDataRow();
+      EXPECT_TRUE(input.isInitialized());
+      output.copyRow(input);
+      output.advanceRow();
+    }
+    return {inputRange.upstreamState(), NoStats{}, output.getClientCall()};
+  };
+  auto forwardCall = [&](AqlItemBlockInputRange& inputRange, OutputAqlItemRow& output)
+      -> std::tuple<ExecutorState, LambdaExe::Stats, AqlCall> {
+    while (inputRange.hasDataRow() && !output.isFull()) {
+      auto const& [state, input] = inputRange.nextDataRow();
+      EXPECT_TRUE(input.isInitialized());
+      output.copyRow(input);
+      output.advanceRow();
+    }
+    return {inputRange.upstreamState(), NoStats{}, output.getClientCall()};
+  };
```

Comment:
Both lambdas look exactly equal to me. Can you just use one?

Reply:
Actually I forgot to call the state asserter, making them slightly different.
I have added refactoring in this file on the shadow row test branch
Here I fixed by calling the asserter